### PR TITLE
feat: useResizeObserver hook

### PR DESCRIPTION
**Pull request by AdrienLemaire**
Recreated the PR #502 using the main branch.

> Inspired by the existing [useMutationObserver](https://react-hooks.org/docs/useMutationObserver), I wrote a similar hook using [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) and thought this might be a useful addition to the library.

### CODE REVIEW

**Comment by imbhargav5** [comment]
Hey @AdrienLemaire the build seems to be failing. Perhaps some types broke.

**Comment by AdrienLemaire** [comment]
@imbhargav5 please consider upgrading the typescript dependency version for rooks.
I cannot see ResizeObserver in 4.0.5, but my project using typescript 4.3.4 is correctly typed:

https://github.com/microsoft/TypeScript/blob/v4.3.4/lib/lib.dom.d.ts#L12515

**Comment by imbhargav5** [comment]
Gotcha. Looks like I have a big weekend in front of me 😂

File: src/hooks/useResizeObserver.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      if (size.height !== height || size.width !== width)
+        setSize(() => ({ height, width }));
+    });
+
+    if (ref.current) observer.observe(ref.current);
+
+    return () => {
+      observer.disconnect();
+    };
+  }, [ref.current]); // eslint-disable-line react-hooks/exhaustive-deps
```
Comment (imbhargav5) [inline src/hooks/useResizeObserver.ts:38]: Using `Ref`.current with `useEffect` won't work. Instead we should use callback refs.

**Review by imbhargav5 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**